src/kms.rs

```diff
@@ -0,0 +1,329 @@
+        let fd = handle.fd;
+        if fd == -1 {
+            return Err(SoftBufferError::IncompleteDisplayHandle);
+        }
+
+        // SAFETY: Invariants guaranteed by the user.
+        let fd = unsafe { BorrowedFd::borrow_raw(fd) };
+
+        Ok(KmsDisplayImpl { fd })
+    }
+}
+
+/// All the necessary types for the Drm/Kms backend.
+#[derive(Debug)]
+pub(crate) struct KmsImpl {
+    /// The display implementation.
+    display: Rc<KmsDisplayImpl>,
+
+    /// The connector to use.
+    connector: connector::Handle,
+
+    /// The CRTC to render to.
+    crtc: crtc::Info,
+
+    /// The dumb buffer we're using as a buffer.
+    buffer: Option<BufferSet>,
+}
+
+/// The buffer implementation.
+pub(crate) struct BufferImpl<'a> {
+    /// The mapping of the dump buffer.
+    mapping: DumbMapping<'a>,
+
+    /// The framebuffer.
+    fb: framebuffer::Handle,
+
+    /// The current size.
+    size: (NonZeroU32, NonZeroU32),
+
+    /// The display implementation.
+    display: &'a KmsDisplayImpl,
+
+    /// The zero buffer.
+    zeroes: &'a [u32],
+}
+
+/// The combined frame buffer and dumb buffer.
+#[derive(Debug)]
+struct BufferSet {
+    /// The frame buffer.
+    fb: framebuffer::Handle,
+
+    /// The dumb buffer.
+    db: DumbBuffer,
+
+    /// Equivalent mapping for reading.
+    zeroes: Box<[u32]>,
+}
+
+impl KmsImpl {
+    /// Create a new KMS backend.
+    ///
+    /// # Safety
+    ///
+    /// The plane must be valid for the lifetime of the backend.
+    pub(crate) unsafe fn new(
+        window_handle: DrmWindowHandle,
+        display: Rc<KmsDisplayImpl>,
+    ) -> Result<Self, SoftBufferError> {
+        log::trace!("new: window_handle={:X}", window_handle.plane);
+
+        // Make sure that the window handle is valid.
+        let plane_handle = match NonZeroU32::new(window_handle.plane) {
+            Some(handle) => plane::Handle::from(handle),
+            None => return Err(SoftBufferError::IncompleteWindowHandle),
+        };
+
+        let plane_info = display
+            .get_plane(plane_handle)
+            .swbuf_err("failed to get plane info")?;
+        let handles = display
+            .resource_handles()
+            .swbuf_err("failed to get resource handles")?;
+
+        // Use either the attached CRTC or the primary CRTC.
+        let crtc = match plane_info.crtc() {
+            Some(crtc) => crtc,
+            None => {
+                log::warn!("no CRTC attached to plane, falling back to primary CRTC");
+                handles
+                    .filter_crtcs(plane_info.possible_crtcs())
+                    .first()
+                    .copied()
+                    .swbuf_err("failed to find a primary CRTC")?
+            }
+        };
+
+        // Use a preferred connector or just select the first one.
+        let connector = handles
+            .connectors
+            .iter()
+            .flat_map(|handle| display.get_connector(*handle, false))
```

Comment:
Hm, it seems strange for softbuffer to be dealing with identifying what connector to use. Should that not be part of the `DrmRawWindowHandle`?

Reply:
It should, I'll file that PR. But it probably won't be released until rwh v0.6, so I think this solution is best for now.

Reply:
Actually nevermind, I found a good solution for this for now.